File: boj/b_17299.js
function solution(n, numbers) {
  const result = Array(n).fill(-1);
  const s = [];
  const cTable = Array(Math.max(...numbers) + 1).fill(0);

  numbers.forEach((number, i) => (cTable[number] += 1));
  numbers.forEach((number, i) => {
    while (s.length && cTable[numbers[s[s.length - 1]]] < cTable[number]) {
      result[s.pop()] = number;
    }
    s.push(i);
  });
  return result;
}
console.log(solution(3, [1000000, 2, 2]));
console.log(solution(7, [1, 1, 2, 3, 4, 2, 1]));

File: boj/b_17299.py
import sys
sys.stdin = open("input.txt", "r") # 제거
input = sys.stdin.readline

n = int(input())
table = list(map(int, input().split()))
s = []
c_table = [0] * (max(table) + 1)
result = [-1 for _ in range(n)]
for i in table:
    c_table[i] += 1
for i in range(n):
    while s and c_table[table[s[-1]]] < c_table[table[i]]:
        result[s.pop()] = table[i]
    else:
        s.append(i)
print(*result)